src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1,487 +1,651 @@
+            self.libvirt_domain_name = f'{self.vm_data["id"]}_{self.vm_data["name"]}'
+            self.__define_domain()
+
+    def status(self):
+        domain = self.domain
+        return {
+            'state': 'STOPPED' if not domain.isActive() else 'RUNNING',
+            'pid': None if not domain.isActive() else self.domain.ID(),
+            'domain_state': DomainState(domain.state()[0]).name,
+        }
+
+    def __define_domain(self):
+        if self.domain:
+            raise CallError(f'{self.libvirt_domain_name} domain has already been defined')
+
+        vm_xml = etree.tostring(self.construct_xml()).decode()
+        if not self.connection.defineXML(vm_xml):
+            raise CallError(f'Unable to define persistent domain for {self.libvirt_domain_name}')
+
+        self.domain = self.connection.lookupByName(self.libvirt_domain_name)
+
+    def undefine_domain(self):
+        if self.domain.isActive():
+            raise CallError(f'Domain {self.libvirt_domain_name} is active. Please stop it first')
+
+        if self.vm_data['bootloader'] == 'GRUB':
+            shutil.rmtree(
+                os.path.join('/tmp/grub', self.libvirt_domain_name), ignore_errors=True
+            )
+
+        self.domain.undefine()
+        self.domain = None
 
+    def __getattribute__(self, item):
+        retrieved_item = object.__getattribute__(self, item)
+        if callable(retrieved_item) and item in ('start', 'stop', 'restart', 'poweroff', 'undefine_domain', 'status'):
+            if not getattr(self, 'domain', None):
+                raise RuntimeError('Domain attribute not defined, please re-instantiate the VM class')
 
-class VMSupervisor(object):
+        return retrieved_item
 
-    def __init__(self, manager, vm):
-        self.manager = manager
+    def start(self, vm_data=None):
+        if self.domain.isActive():
+            raise CallError(f'{self.libvirt_domain_name} domain is already active')
 
-        os.makedirs('/var/log/vm', exist_ok=True)
-        self.logger = self.manager.logger.getChild(f'vm_{vm["id"]}')
+        self.update_domain(vm_data)
 
-        handler = middlewared.logger.ErrorProneRotatingFileHandler(
-            f'/var/log/vm/{vm["name"]}_{vm["id"]}',
-            maxBytes=10485760,
-            backupCount=5,
+        # Let's ensure that we are able to boot a GRUB based VM
+        if self.vm_data['bootloader'] == 'GRUB' and not any(
+            isinstance(d, RAW) and d.data['attributes'].get('boot') for d in self.devices
+        ):
+            raise CallError(f'Unable to find boot devices for {self.libvirt_domain_name} domain')
+
+        if len([d for d in self.devices if isinstance(d, VNC)]) > 1:
+            raise CallError('Only one VNC device per VM is supported')
+
+        for device in self.devices:
+            device.pre_start_vm()
+
+        if self.domain.create() < 0:
+            raise CallError(f'Failed to boot {self.vm_data["name"]} domain')
```

Comment:
Changes made by `pre_start_vm` won't be rolled back if error happens

Reply:
Should they be ?